### PR TITLE
Fixes for #648 and #357

### PR DESCRIPTION
**Pull request by tkvangorder**
This pull request addresses two issues that we encountered as we moved our code base to use Spring Cloud.

https://github.com/spring-cloud/spring-cloud-netflix/issues/648

The primary fix is to add retry capabilities into the RestTemplate such that it will respect the Ribbon Retry settings similar to how the Feign client works.

This fix changes the RibbonHttpRequest to use the 'executeWithLoadBalancer()'. This change leverages Ribbon's LoadBalancedCommand to issue the rest call. This command will correctly select a server instance and update the server stats for that server. Additionally, this command will also retry the request should a connection failure occur. Because this logic is done in the Ribbon layer, it is no longer necessary to choose a server instance in the RibbonClientHttpRequestFactory or to collect the serverstats in the RibbonHttpRequest.

There is a new unit test, `RestTemplateRetryTest` that exercises the logic and insures the server stats are being correctly updated and that retry logic is working as expected. 

Notes: 

I ran into an issue where when a Ribbon client is configured without the use of Eureka, it's 
'DeploymentContextBasedVipAddresses' property was not set. Failure to set this means the ribbon client will not attempt to resolve the client ID to one of the server instances. I found code in 'EurekaRibbonClientConfiguration' that was setting this and I ended up copying this code to RibbonClientConfiguration. This may not be the best way to solve this, @spencergibb or @dsyer is there a better place to do something like this? 

The unit test leverages a simple AvailabilityFilteringRule that just skips servers that have open circuit breakers. It is more difficult to come up with reproducible, easy checks for other types of rules...although I used this test to figure out how other Rules actually work.

This change will affect anyone that is wrapping their rest calls in a high level "retry" mechanism (such as hystrix or using Spring's Retry annotation.

https://github.com/spring-cloud/spring-cloud-netflix/issues/357 

This issue was closed by the original author because he had hacked around the problem. In our case we had a large code base that made this approach challenging. Our fix was again to rely on the Ribbon library to correctly manipulate the Http header. 

See my comment at the end of this issue for details.

### CODE REVIEW

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientConfiguration.java [hunk 1]
```diff
@@ -155,6 +166,28 @@ public ServerIntrospector serverIntrospector() {
 		return new DefaultServerIntrospector();
 	}
 
+	@PostConstruct
+	public void preprocess() {
+		setProp(name, DeploymentContextBasedVipAddresses.key(), name);
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientConfiguration.java]: Why did you need to copy this from `EurekaRibbonClientConfiguration`?

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientConfiguration.java]: And now I re-read your comments :-) Now I wonder why it's needed in both places.

Comment (tkvangorder) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientConfiguration.java]: Yeah, It is definitely redundant...I dont know enough about the internals of how Eureka is used to create a ribbon client dynamically vs. having a ribbon client explicitly defined with RibbonClientConfiguration.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java [hunk 2]
```diff
@@ -38,12 +34,8 @@
 
 	private final SpringClientFactory clientFactory;
 
-	private LoadBalancerClient loadBalancer;
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java]: This really worries me, though I can't put specifics on it yet.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java]: Looking at `RibbonLoadBalancerClient` now.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java]: Probably the biggest thing I can see is automatically detecting `isSecure` vs. having to set the value in properties.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java]: Looks like that might be taken care of with `OverrideRestClient`.

**Comment by spencergibb** [comment]
@dsyer thoughts?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java (hunk 2 above)
Comment (tkvangorder) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java]: yeah, its a little confusing because the LoadBalancerCommand is doing similar things to the logic in the RibbonLoadBalancerClient. Is the "isSecure" flag there so that if some instances are secure and some are not, the URI will have the correct "http" or "https" when using that server?

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java]: Yeah. We plug into the `RestClient` here https://github.com/spring-cloud/spring-cloud-netflix/blob/master/spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientConfiguration.java#L170-L177

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/ribbon/RibbonClientHttpRequestFactory.java]: I'm trying to remember why we did that in the first place. I'm thinking it was to use the `LoadBalancerClient` abstraction, but at that point, we're in the middle of ribbon already.

**Comment by spencergibb** [comment]
@tkvangorder after talking with @dsyer we'd like this feature to be opt-in via a config property. Is that something you'd be willing to change?

**Comment by spencergibb** [comment]
Actually, I'm in the middle of doing #961 which makes `RibbonClientHttpRequestFactory` opt-in. Since that is where your changes are, #961 effectively makes this opt-in. Once I merge, can you rebase?

**Comment by spencergibb** [comment]
Actually, I can rebase.

**Comment by tkvangorder** [comment]
Haha, I dont even time to respond. 8) So, if I am understanding this correctly, after your changes:

Using The RibbonClientHttpRequestFactory (which uses Netflix's RestClient) will be an opt-in. 
And if you opt-in, you will also get the retry logic from my pull request?

**Comment by spencergibb** [comment]
Yes, that is what I'm saying.

**Comment by spencergibb** [comment]
Squashed and merged with some polish via 4479a2b2ee28cba1ea86dcca18805cce369634d8

**Comment by tkvangorder** [comment]
Awesome! It looks like this will get into Brixton?

**Comment by spencergibb** [comment]
Yes, I wanted it before RC2 which should be in the next week.
